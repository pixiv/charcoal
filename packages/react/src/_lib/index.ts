/**
 * 今後ポートされる予定の汎用的な関数群
 */

/**
 * Function used to assert a given code path is unreachable
 */
export function unreachable(): never
/**
 * Function used to assert a given code path is unreachable.
 * Very useful for ensuring switches are exhaustive:
 *
 * ```ts
 * switch (a.type) {
 *   case Types.A:
 *   case Types.B:
 *     break
 *   default:
 *     unreachable(a) // will cause a build error if there was
 *                    // a Types.C that was not checked
 * }
 * ```
 *
 * @param value Value to be asserted as unreachable
 */
// NOTE: Uses separate overloads, _not_ `value?: never`, to not allow `undefined` to be passed

export function unreachable(value: never): never
export function unreachable(value?: never): never {
  throw new Error(
    arguments.length === 0
      ? 'unreachable'
      : `unreachable (${JSON.stringify(value)})`
  )
}

/**
 * 複数のrefをマージする。
 *
 * forwardRefで受け取ったrefと、コンポーネント内で定義したrefを同じ要素につけたいケースなどで使う
 */
export function mergeRefs<T>(...refs: React.Ref<T>[]): React.RefCallback<T> {
  return (value) => {
    for (const ref of refs) {
      if (typeof ref === 'function') {
        ref(value)
      } else if (ref !== null) {
        // TODO: use RefObject
        // https://react.dev/blog/2024/04/25/react-19-upgrade-guide#:~:text=%3D%201%3B-,MutableRef,-is%20now%20deprecated
        ;(ref as React.MutableRefObject<T | null>).current = value
      }
    }
  }
}

export function countCodePointsInString(string: string): number {
  // [...string] とするとproduction buildで動かなくなる
  // cf. https://twitter.com/f_subal/status/1497214727511891972
  return Array.from(string).length
}
