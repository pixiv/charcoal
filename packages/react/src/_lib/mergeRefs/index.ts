import React from 'react'

/**
 * Custom implementation of mergeRefs to replace @react-aria/utils
 * Merges multiple refs into a single ref callback
 */
export function mergeRefs<T>(
  ...refs: (React.Ref<T> | undefined)[]
): React.RefCallback<T> {
  return (value: T | null) => {
    refs.forEach((ref) => {
      if (ref == null) {
        return
      }

      if (typeof ref === 'function') {
        ref(value)
      } else if (typeof ref === 'object') {
        ;(ref as React.MutableRefObject<T | null>).current = value
      }
    })
  }
}
