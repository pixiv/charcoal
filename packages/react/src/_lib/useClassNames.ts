import { useMemo } from 'react'

/**
 * Join some class names if propsClassName is defined.
 */
export function useClassNames(...classNames: (string | undefined)[]): string {
  return useMemo(() => classNames.filter((v) => v).join(' '), [classNames])
}
