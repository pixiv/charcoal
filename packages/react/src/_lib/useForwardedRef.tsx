import * as React from 'react'

export function useForwardedRef<T>(ref: React.ForwardedRef<T>): React.RefObject<T> {
  const innerRef = React.useRef<T>(null)

  React.useEffect(() => {
    if (!ref) return
    if (typeof ref === 'function') {
      ref(innerRef.current)
    } else {
      ref.current = innerRef.current
    }
  })

  return innerRef
}
