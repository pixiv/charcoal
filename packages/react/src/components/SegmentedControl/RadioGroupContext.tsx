import React, { createContext, useContext } from 'react'
import { RadioGroupState } from 'react-stately'

const RadioContext = createContext<RadioGroupState | null>(null)

type RadioProviderProps = React.PropsWithChildren<{
  value: RadioGroupState
}>
export const RadioProvider: React.FC<RadioProviderProps> = ({
  value,
  children,
}) => {
  return <RadioContext.Provider value={value}>{children}</RadioContext.Provider>
}
export const useRadioContext = () => {
  const state = useContext(RadioContext)

  if (state === null)
    throw new Error('`<RadioProvider>` is not likely mounted.')

  return state
}
