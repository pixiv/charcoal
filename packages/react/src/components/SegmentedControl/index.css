.charcoal-segmented-control {
  display: inline-flex;
  align-items: center;

  background-color: var(--charcoal-color-container-secondary-default-a);
  border-radius: 16px;
}

.charcoal-segmented-control-radio__label {
  position: relative;
  display: flex;
  align-items: center;
  cursor: pointer;
  height: 32px;

  padding-right: 16px;
  padding-left: 16px;
  border-radius: 16px;
  color: var(--charcoal-color-text-secondary-default);

  font-size: 14px;
  line-height: 22px;
}

.charcoal-segmented-control-radio__label[aria-disabled]:not(
    [aria-disabled='false']
  ) {
  cursor: default;
  opacity: 0.32;
}

.charcoal-segmented-control-radio__label[data-checked='true'] {
  background-color: var(--charcoal-color-container-primary-default);
  color: var(--charcoal-color-text-on-primary-default);
}

.charcoal-segmented-control-radio__input {
  position: absolute;

  height: 0px;
  width: 0px;
  padding: 0;
  margin: 0;

  appearance: none;
  box-sizing: border-box;
  overflow: hidden;
  white-space: nowrap;
  opacity: 0;
}
