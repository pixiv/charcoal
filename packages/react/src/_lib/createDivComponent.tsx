import { DetailedHTMLProps, DO_NOT_USE_OR_YOU_WILL_BE_FIRED_CALLBACK_REF_RETURN_VALUES, forwardRef, ForwardRefExoticComponent, HTMLAttributes, RefAttributes, RefObject } from 'react'
import { useClassNames } from './useClassNames'

export function createDivComponent(mainClassName: string): ForwardRefExoticComponent<Omit<Omit<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, "ref"> & {
  ref?: ((instance: HTMLDivElement | null) => void | DO_NOT_USE_OR_YOU_WILL_BE_FIRED_CALLBACK_REF_RETURN_VALUES[keyof DO_NOT_USE_OR_YOU_WILL_BE_FIRED_CALLBACK_REF_RETURN_VALUES]) | RefObject<HTMLDivElement> | null | undefined
}, "ref"> & RefAttributes<HTMLDivElement>> {
  return forwardRef<HTMLDivElement, React.ComponentPropsWithRef<'div'>>(
    function DivComponent({ className, ...props }, ref) {
      const classNames = useClassNames(mainClassName, className)
      return <div className={classNames} ref={ref} {...props} />
    }
  )
}
