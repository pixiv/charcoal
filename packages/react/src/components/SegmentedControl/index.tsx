import React, { ReactNode, forwardRef, memo, useMemo, useRef } from 'react'
import { useRadioGroupState } from 'react-stately'
import {
  AriaRadioGroupProps,
  AriaRadioProps,
  useRadio,
  useRadioGroup,
} from '@react-aria/radio'
import styled from 'styled-components'
import { disabledSelector } from '@charcoal-ui/utils'

import { RadioProvider, useRadioContext } from './RadioGroupContext'
import { theme } from '../../styled'

type SegmentedControlItem = {
  label: React.ReactNode
  value: string
  disabled?: boolean
}

export type SegmentedControlProps = {
  readonly id?: string
  readonly name?: string
  readonly disabled?: boolean
  readonly readonly?: boolean
  readonly required?: boolean

  readonly value?: string
  readonly defaultValue?: string

  readonly data: string[] | SegmentedControlItem[]

  readonly onChange?: (value: string) => void
}

const SegmentedControl = forwardRef<HTMLDivElement, SegmentedControlProps>(
  function SegmentedControlInner(props, ref) {
    const ariaRadioGroupProps = useMemo<AriaRadioGroupProps>(
      () => ({
        ...props,
        isDisabled: props.disabled,
        isReadOnly: props.readonly,
        isRequired: props.required,
        'aria-label': props.name,
      }),
      [props]
    )
    const state = useRadioGroupState(ariaRadioGroupProps)
    const { radioGroupProps } = useRadioGroup(ariaRadioGroupProps, state)
    const segmentedControlItems = useMemo<SegmentedControlItem[]>(() => {
      return props.data.map((d) =>
        typeof d === 'string' ? { value: d, label: d } : d
      )
    }, [props.data])

    return (
      <SegmentedControlRoot ref={ref} {...radioGroupProps}>
        <RadioProvider value={state}>
          {segmentedControlItems.map((item) => (
            <Segmented
              key={item.value}
              value={item.value}
              disabled={item.disabled}
            >
              {item.label}
            </Segmented>
          ))}
        </RadioProvider>
      </SegmentedControlRoot>
    )
  }
)

export default memo(SegmentedControl)

type RadioProps = {
  value: string
  disabled?: boolean
  children?: ReactNode
}

const Segmented = (props: RadioProps) => {
  const state = useRadioContext()
  const ref = useRef<HTMLInputElement>(null)
  const ariaRadioProps = useMemo<AriaRadioProps>(
    () => ({
      value: props.value,
      isDisabled: props.disabled,
      children: props.children,
    }),
    [props]
  )

  const { inputProps, isDisabled, isSelected } = useRadio(
    ariaRadioProps,
    state,
    ref
  )

  return (
    <SegmentedRoot
      aria-disabled={isDisabled || state.isReadOnly}
      checked={isSelected}
    >
      <SegmentedInput {...inputProps} ref={ref} />
      <RadioLabel>
        <SegmentedLabelInner>{props.children}</SegmentedLabelInner>
      </RadioLabel>
    </SegmentedRoot>
  )
}

const SegmentedControlRoot = styled.div`
  display: inline-flex;
  align-items: center;

  ${theme((o) => [o.bg.surface3, o.borderRadius(16)])}
`

const SegmentedRoot = styled.label<{ checked?: boolean }>`
  position: relative;
  display: flex;
  align-items: center;
  cursor: pointer;
  height: 32px;

  ${disabledSelector} {
    cursor: default;
  }

  ${({ checked }) =>
    theme((o) => [
      o.padding.horizontal(16),
      o.borderRadius(16),
      o.disabled,
      checked === true && o.bg.brand,
      checked === true ? o.font.text5 : o.font.text2,
    ])}
`
const SegmentedInput = styled.input`
  position: absolute;

  height: 0px;
  width: 0px;
  padding: 0;
  margin: 0;

  appearance: none;
  box-sizing: border-box;
  overflow: hidden;
  white-space: nowrap;
  opacity: 0;
`

const RadioLabel = styled.div`
  background: transparent;
  display: flex;
  align-items: center;
  height: 22px;
`
const SegmentedLabelInner = styled.div`
  ${theme((o) => [o.typography(14)])}
`
