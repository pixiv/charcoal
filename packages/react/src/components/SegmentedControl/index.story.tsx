import { action } from '@storybook/addon-actions'
import React from 'react'
import SegmentedControl, { SegmentedControlProps } from '.'
import { Story } from '../../_lib/compat'

export default {
  title: 'SegmentedControl',
  component: SegmentedControl,
}

export const StringSegments: Story<SegmentedControlProps> = (props) => {
  return <SegmentedControl {...props} onChange={action('change')} />
}

StringSegments.args = {
  data: ['option1', 'option2', 'option3'],
  disabled: false,
  readonly: false,
  required: false,
}

export const ObjectSegments: Story<SegmentedControlProps> = (props) => {
  return <SegmentedControl {...props} onChange={action('change')} />
}

ObjectSegments.args = {
  data: [
    { label: '選択肢1', value: 'option1' },
    { label: '選択肢2', value: 'option2' },
    { label: '選択肢3', value: 'option3' },
    { label: '選択肢4', value: 'option4', disabled: true },
  ],
  disabled: false,
  readonly: false,
  required: false,
}
