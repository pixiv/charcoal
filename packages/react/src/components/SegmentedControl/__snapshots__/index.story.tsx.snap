// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots SegmentedControl Object Segments 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: var(--charcoal-surface3);
  border-radius: 16px;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  height: 32px;
  padding-right: 16px;
  padding-left: 16px;
  border-radius: 16px;
  color: var(--charcoal-text2);
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled='false']) {
  cursor: default;
  opacity: 0.32;
}

.c2 {
  position: absolute;
  height: 0px;
  width: 0px;
  padding: 0;
  margin: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  box-sizing: border-box;
  overflow: hidden;
  white-space: nowrap;
  opacity: 0;
}

.c3 {
  background: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 22px;
}

.c4 {
  font-size: 14px;
  line-height: 22px;
  display: flow-root;
}

.c4::before {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-top: -4px;
}

.c4::after {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-bottom: -4px;
}

<div
  data-dark={false}
>
  <div
    aria-label="test"
    aria-orientation="vertical"
    className="c0"
    id="react-aria-1"
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    role="radiogroup"
  >
    <label
      aria-disabled={false}
      checked={false}
      className="c1"
    >
      <input
        checked={false}
        className="c2"
        disabled={false}
        name="test"
        onChange={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        required={false}
        tabIndex={0}
        type="radio"
        value="option1"
      />
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          選択肢1
        </div>
      </div>
    </label>
    <label
      aria-disabled={false}
      checked={false}
      className="c1"
    >
      <input
        checked={false}
        className="c2"
        disabled={false}
        name="test"
        onChange={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        required={false}
        tabIndex={0}
        type="radio"
        value="option2"
      />
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          選択肢2
        </div>
      </div>
    </label>
    <label
      aria-disabled={false}
      checked={false}
      className="c1"
    >
      <input
        checked={false}
        className="c2"
        disabled={false}
        name="test"
        onChange={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        required={false}
        tabIndex={0}
        type="radio"
        value="option3"
      />
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          選択肢3
        </div>
      </div>
    </label>
    <label
      aria-disabled={true}
      checked={false}
      className="c1"
    >
      <input
        checked={false}
        className="c2"
        disabled={true}
        name="test"
        onChange={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        required={false}
        type="radio"
        value="option4"
      />
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          選択肢4
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`Storyshots SegmentedControl String Segments 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: var(--charcoal-surface3);
  border-radius: 16px;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  height: 32px;
  padding-right: 16px;
  padding-left: 16px;
  border-radius: 16px;
  color: var(--charcoal-text2);
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled='false']) {
  cursor: default;
  opacity: 0.32;
}

.c2 {
  position: absolute;
  height: 0px;
  width: 0px;
  padding: 0;
  margin: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  box-sizing: border-box;
  overflow: hidden;
  white-space: nowrap;
  opacity: 0;
}

.c3 {
  background: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 22px;
}

.c4 {
  font-size: 14px;
  line-height: 22px;
  display: flow-root;
}

.c4::before {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-top: -4px;
}

.c4::after {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-bottom: -4px;
}

<div
  data-dark={false}
>
  <div
    aria-label="test"
    aria-orientation="vertical"
    className="c0"
    id="react-aria-5"
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    role="radiogroup"
  >
    <label
      aria-disabled={false}
      checked={false}
      className="c1"
    >
      <input
        checked={false}
        className="c2"
        disabled={false}
        name="test"
        onChange={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        required={false}
        tabIndex={0}
        type="radio"
        value="option1"
      />
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          option1
        </div>
      </div>
    </label>
    <label
      aria-disabled={false}
      checked={false}
      className="c1"
    >
      <input
        checked={false}
        className="c2"
        disabled={false}
        name="test"
        onChange={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        required={false}
        tabIndex={0}
        type="radio"
        value="option2"
      />
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          option2
        </div>
      </div>
    </label>
    <label
      aria-disabled={false}
      checked={false}
      className="c1"
    >
      <input
        checked={false}
        className="c2"
        disabled={false}
        name="test"
        onChange={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        required={false}
        tabIndex={0}
        type="radio"
        value="option3"
      />
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          option3
        </div>
      </div>
    </label>
  </div>
</div>
`;
